# Hola Yair!

Mi nombre es David Bautista, soy code reviewer de Practicum y voy a revisar el proyecto que acabas de desarrollar.

Cuando vea un error la primera vez, lo señalaré. Deberás encontrarlo y arreglarlo. La intención es que te prepares para un espacio real de trabajo. En un trabajo, el líder de tu equipo hará lo mismo. Si no puedes solucionar el error, te daré más información en la próxima ocasión.

Encontrarás mis comentarios más abajo - por favor, no los muevas, no los modifiques ni los borres.

¿Cómo lo voy a hacer? Voy a leer detenidamente cada una de las implementaciones que has llevado a cabo para cumplir con lo solicitado. Verás los comentarios de esta forma:

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Si todo está perfecto.
</div>


<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Si tu código está bien pero se puede mejorar o hay algún detalle que le hace falta.
</div>


<div class="alert alert-block alert-danger">
    
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
    
Si de pronto hace falta algo o existe algún problema con tu código o conclusiones.
</div>


Puedes responderme de esta forma: 

<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class="tocSkip"></a>
</div>

¡Empecemos!


# Déjame escuchar la música

# Contenido <a id='back'></a>

* [Introducción](#intro)
* [Etapa 1. Descripción de los datos](#data_review)
    * [Conclusiones](#data_review_conclusions)
* [Etapa 2. Preprocesamiento de datos](#data_preprocessing)
    * [2.1 Estilo del encabezado](#header_style)
    * [2.2 Valores ausentes](#missing_values)
    * [2.3 Duplicados](#duplicates)
    * [2.4 Conclusiones](#data_preprocessing_conclusions)
* [Etapa 3. Prueba de hipótesis](#hypotheses)
    * [3.1 Hipótesis 1: actividad de los usuarios y las usuarias en las dos ciudades](#activity)
    * [3.2 Hipótesis 2: preferencias musicales los lunes y los viernes](#week)
    * [3.3 Hipótesis 3: preferencias de género en Springfield y Shelbyville](#genre)
* [Conclusiones](#end)

## Introducción <a id='intro'></a>
Como analista de datos, tu trabajo consiste en analizar datos para extraer información valiosa y tomar decisiones basadas en datos. Esto implica diferentes etapas, como la descripción general de los datos, el preprocesamiento y la prueba de hipótesis.

Siempre que investigamos, necesitamos formular hipótesis que después podamos probar. A veces aceptamos estas hipótesis; otras, las rechazamos. Para tomar las decisiones correctas, una empresa debe ser capaz de entender si está haciendo las suposiciones correctas.

En este proyecto, compararás las preferencias musicales de las ciudades de Springfield y Shelbyville. Estudiarás datos reales de transmisión de música online para probar las hipótesis a continuación y comparar el comportamiento de los usuarios y las usuarias de estas dos ciudades.

### Objetivo:
Prueba tres hipótesis:
1. La actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la cuidad.
2. Los lunes por la mañana, los habitantes de Springfield y Shelbyville escuchan géneros distintos. Lo mismo ocurre con los viernes por la noche.
3. Los oyentes de Springfield y Shelbyville tienen preferencias distintas. En Springfield prefieren el pop, mientras que en Shelbyville hay más personas a las que les gusta el rap.

### Etapas
Los datos del comportamiento del usuario se almacenan en el archivo `/datasets/music_project_en.csv`. No hay ninguna información sobre la calidad de los datos, así que necesitarás examinarlos antes de probar las hipótesis.

Primero, evaluarás la calidad de los datos y verás si los problemas son significativos. Entonces, durante el preprocesamiento de datos, tomarás en cuenta los problemas más críticos.

Tu proyecto consistirá en tres etapas:
 1. Descripción de los datos
 2. Preprocesamiento de datos
 3. Prueba de hipótesis


### Reto

En este proyecto, preparamos un pequeño reto para ti. Incluimos un nuevo tipo de estructura de datos: las marcas temporales. Las marcas temporales son muy comunes y merecen una atención adicional. Más adelante en el programa, aprenderás mucho sobre ellas. Sin embargo, por ahora las trataremos como simples strings. Necesitamos marcas temporales en este proyecto para poner a prueba una de nuestras hipótesis. No te preocupes, te ayudaremos con esto. Tu nivel de conocimientos actual será suficiente para abordarlo.

Por ejemplo, digamos que tenemos dos marcas temporales: `dt1 = "12:00:00"` y `dt2 = "06:00:00"`. Queremos comparar estas dos marcas temporales y ver cuál es posterior.

Podemos compararlas mediante los operadores de comparación estándar (`<`, `>`, `<=`, `>=`, `==`, `!=`). Ejecuta la siguiente celda de código para comparar dos marcas temporales:


In [ ]:
# Comparar los objetos datetime

dt1 = "12:00:00"
dt2 = "06:00:00"

if dt1 < dt2:
    print("La marca temporal 2 es posterior")
else:
    print("La marca temporal 1 es posterior")

Timestamp 1 is later


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Muy bien, recuerda que para todo los proyectos que realices es importante genera un contexto donde se comente que trata el caso y cuáles con los objetivos a cumplir, además es indispensable que se genere una tabla de contenido con el fin de mantener el orden en la entrega, Por otro lado, buen trabajo con el reto.

[Volver a Contenidos](#back)

## Etapa 1. Descripción de los datos <a id='data_review'></a>

Abre los datos y examínalos.

Necesitarás `pandas`, así que impórtalo.

In [2]:
# importar pandas
import pandas as pd

In [3]:
pd.read_csv("/datasets/music_project_en.csv")

userID                              Track            artist  \
0      FFB692EC                  Kamigata To Boots  The Mass Missile   
1      55204538        Delayed Because of Accident  Andreas Rönnberg   
2        20EC38                  Funiculì funiculà       Mario Lanza   
3      A3DD03C9              Dragons in the Sunset        Fire + Ice   
4      E2DC1FAE                        Soul People        Space Echo   
...         ...                                ...               ...   
65074  729CBB09                            My Name            McLean   
65075  D08D4A55  Maybe One Day (feat. Black Spade)       Blu & Exile   
65076  C5E3A0D5                          Jalopiina               NaN   
65077  321D0506                      Freight Train     Chas McDevitt   
65078  3A64EF84          Tell Me Sweet Little Lies      Monica Lopez   

            genre       City        time        Day  
0            rock  Shelbyville  20:28:33  Wednesday  
1            rock  Springfield  14:07:09     Friday  
2             pop  Shelbyville  20:58:07  Wednesday  
3            folk  Shelbyville  08:37:09     Monday  
4           dance  Springfield  08:34:34     Monday  
...           ...          ...       ...        ...  
65074         rnb  Springfield  13:32:28  Wednesday  
65075         hip  Shelbyville  10:00:00     Monday  
65076  industrial  Springfield  20:09:26     Friday  
65077        rock  Springfield  21:43:59     Friday  
65078     country  Springfield  21:59:46     Friday  

[65079 rows x 7 columns]

Si estás trabajando en la plataforma, lee el archivo `music_project_en.csv` de la carpeta /datasets/ y guárdalo en la variable `df`. Si estás trabajando localmente, no es necesario especificar carpeta alguna siempre y cuando el archivo `music_project_en.csv` esté en la misma carpeta que este Notebook:

In [4]:
# leer el archivo y almacenarlo en df
df = pd.read_csv("/datasets/music_project_en.csv")

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Muy bien.

Muestra las 10 primeras filas de la tabla:

In [5]:
# obtener las 10 primeras filas de la tabla df
display(df.head(10))

userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5  Shelbyville  13:09:41     Friday  
6  Springfield  13:00:07  Wednesday  
7  Springfield  20:47:49  Wednesday  
8  Springfield  09:17:40     Friday  
9  Shelbyville  21:20:49  Wednesday

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
~~Te recomiendo usar ``display()`` en vez de ``print()`` para que las salidas de las tablas sean más estéticas.~~

<div class="alert alert-block alert-success">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>
    
Muy bien.

Obtén la información general sobre la tabla con un comando:

In [7]:
# obtener información general sobre los datos en df
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


<div class="alert alert-block alert-danger">
<strong>Comentario del revisor</strong> <a class="tocSkip"></a>
    
~~Recuerda que para obtener la información del dataframe, debes utilizar el método ``df.info()``.~~
</div>


<div class="alert alert-block alert-success">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>
    
Buen trabajo.

Estas son nuestras observaciones sobre la tabla. Contiene siete columnas. Todas almacenan el mismo tipo de datos: `object` (objeto).

Según la documentación:
- `'userID'` — identificador del usuario o la usuaria;
- `'Track'` — título de la canción;
- `'artist'` — nombre del artista;
- `'genre'` — género musical;
- `'City'` — ciudad del usuario o la usuaria;
- `'time'` — hora exacta en la que se reprodujo la pista;
- `'Day'` — día de la semana.

Podemos ver tres problemas con el estilo en los encabezados de la tabla:
1. Algunos encabezados están en mayúsculas, otros, en minúsculas.
2. Hay espacios en algunos encabezados.
3. `Detecta el tercer problema por tu cuenta y descríbelo aquí`. hay q cambiar userid por user_id




### Tus observaciones <a id='data_review_conclusions'></a>

`Escriba sus observaciones aquí:`

`1.   ¿Qué tipo de datos tenemos a nuestra disposición en las filas? ¿Y cómo podemos entender lo que almacenan las columnas?`
Los tipos de datos son object y lo que podemos ver en las columnas estan ordenados segun su clasificacion

`2.   ¿Hay suficientes datos para proporcionar respuestas a nuestras tres hipótesis, o necesitamos más información?`
si tenemos todos los datos necesarios

`3.   ¿Notaste algún problema en los datos, como valores ausentes, duplicados o tipos de datos incorrectos?`
segun la vista pude notar algunos datos ausentes 

[Volver a Contenidos](#back)

## Etapa 2. Preprocesamiento de datos <a id='data_preprocessing'></a>

El objetivo aquí es preparar los datos para que sean analizados.
El primer paso es resolver cualquier problema con los encabezados. Luego podemos avanzar a los valores ausentes y duplicados. Empecemos.

Corrige el formato en los encabezados de la tabla.


### Estilo del encabezado <a id='header_style'></a>
Muestra los encabezados de la tabla:

In [8]:
# la lista de encabezados para la tabla df
print(df.columns)

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')


Cambia los encabezados de la tabla de acuerdo con las reglas del buen estilo:
* todos los caracteres deben ser minúsculas;
* elimina los espacios;
* si el nombre tiene varias palabras, utiliza snake_case.

Pon todos los caracteres en minúsculas e imprime el encabezado de la tabla de nuevo:

In [9]:
# bucle en los encabezados poniendo todo en minúsculas
new_column = []
for minu in df.columns:
    minus = minu.lower()
    new_column.append(minus)
print(new_column)    

['  userid', 'track', 'artist', 'genre', '  city  ', 'time', 'day']


Ahora elimina los espacios al principio y al final de los encabezados y muéstralos:

In [10]:
# bucle en los encabezados eliminando los espacios
new_column=[]
for minu in df.columns:
    minus = minu.lower()
    espaces = minus.strip()
    print(espaces)

userid
track
artist
genre
city
time
day


Aplica snake_case al encabezado userID e imprime el encabezado de la tabla:

In [11]:
# cambiar el nombre del encabezado "user_id"
new_column=[]
for minu in df.columns:
    minus = minu.lower()
    espaces = minus.strip()
    snake = espaces.replace("userid","user_id")
    print(snake)   

user_id
track
artist
genre
city
time
day


Comprueba el resultado. Muestra los encabezados una vez más:

In [12]:
# comprobar el resultado: la lista de encabezados
new_column=[]
for minu in df.columns:
    minus = minu.lower()
    espaces = minus.strip()
    snake = espaces.replace("userid","user_id")
    new_column.append(snake)
df.columns = new_column
print(new_column)  

['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day']


[Volver a Contenidos](#back)

<div class="alert alert-block alert-danger">
<strong>Comentario del revisor</strong> <a class="tocSkip"></a>
    
~~Estás explorando el problema de diferentes maneras, sin embargo, lo ideal es que te quedes con la versión final (que es la correcta) y elimines las otras opciones.~~
</div>

<div class="alert alert-block alert-info">
<b>Respuesta del estudiante hice las otras versiones incompletas porque asi me lo pedia la guia </b> <a class="tocSkip"></a>
</div>

<div class="alert alert-block alert-success">
<strong>Comentario del revisor #2</strong> <a class="tocSkip"></a>
    
Entiendo!. Buen trabajo.
</div>


### Valores ausentes <a id='missing_values'></a>
Primero, encuentra el número de valores ausentes en la tabla. Para ello, utiliza dos métodos `pandas`:

In [13]:
# calcular el número de valores ausentes
valores_ausentes = df.isna().sum()
print(valores_ausentes)

user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64


No todos los valores ausentes afectan a la investigación. Por ejemplo, los valores ausentes en `track` y `artist` no son cruciales. Simplemente puedes reemplazarlos con valores predeterminados como el string `'unknown'` (desconocido).

Pero los valores ausentes en `'genre'` pueden afectar la comparación entre las preferencias musicales de Springfield y Shelbyville. En la vida real, sería útil saber las razones por las cuales hay datos ausentes e intentar recuperarlos. Pero no tenemos esa oportunidad en este proyecto. Así que tendrás que:
* rellenar estos valores ausentes con un valor predeterminado;
* evaluar cuánto podrían afectar los valores ausentes a tus cómputos;

Reemplazar los valores ausentes en `'track'`, `'artist'` y `'genre'` con el string `'unknown'`. Para hacer esto, crea la lista `columns_to_replace`, recorre sobre ella con un bucle `for`, y para cada columna reemplaza los valores ausentes en ella:

In [14]:
# bucle en los encabezados reemplazando los valores ausentes con 'unknown'
columns_to_replace = ["track", "artist", "genre"]
for correction in columns_to_replace:
    df[correction].fillna("unknow",inplace=True) 



Asegúrate de que la tabla no contiene más valores ausentes. Cuenta de nuevo los valores ausentes.

In [15]:
# contando valores ausentes
valores_ausentes = df.isna().sum()
print(valores_ausentes)

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64


<div class="alert alert-block alert-success">
<strong>Comentario del revisor</strong> <a class="tocSkip"></a>
    
Buen trabajo con los valores nulos.
</div>


[Volver a Contenidos](#back)

### Duplicados <a id='duplicates'></a>
Encuentra el número de duplicados explícitos en la tabla usando un comando:

In [16]:
# contar duplicados explícitos
print(df.duplicated().sum())

3826


Llama al método `pandas` para deshacerte de los duplicados explícitos:

In [17]:
# eliminar duplicados explícitos
df.drop_duplicates(inplace=True)
df.reset_index(drop=True, inplace=True)

Cuenta los duplicados explícitos una vez más para asegurarte de haberlos eliminado todos:

In [18]:
# comprobación de duplicados
print(df.duplicated().sum())

0


<div class="alert alert-block alert-success">
<strong>Comentario del revisor</strong> <a class="tocSkip"></a>
    
Buen trabajo con los valores duplicados.
</div>


Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.

Para hacerlo, primero imprimamos una lista de nombres de género únicos, ordenados en orden alfabético. Para hacerlo:
* recupera la columna deseada del dataFrame;
* llama al método que te devolverá todos los valores de columna únicos;
* aplica un método de ordenamiento a tu resultado.


In [28]:
# inspeccionar los nombres de género únicos
print(df["genre"].sort_values().unique())
print(df["genre"].nunique())

['acid' 'acoustic' 'action' 'adult' 'africa' 'afrikaans' 'alternative'
 'ambient' 'americana' 'animated' 'anime' 'arabesk' 'arabic' 'arena'
 'argentinetango' 'art' 'audiobook' 'avantgarde' 'axé' 'baile' 'balkan'
 'beats' 'bigroom' 'black' 'bluegrass' 'blues' 'bollywood' 'bossa'
 'brazilian' 'breakbeat' 'breaks' 'broadway' 'cantautori' 'cantopop'
 'canzone' 'caribbean' 'caucasian' 'celtic' 'chamber' 'children' 'chill'
 'chinese' 'choral' 'christian' 'christmas' 'classical' 'classicmetal'
 'club' 'colombian' 'comedy' 'conjazz' 'contemporary' 'country' 'cuban'
 'dance' 'dancehall' 'dancepop' 'dark' 'death' 'deep' 'deutschrock'
 'deutschspr' 'dirty' 'disco' 'dnb' 'documentary' 'downbeat' 'downtempo'
 'drum' 'dub' 'dubstep' 'eastern' 'easy' 'electronic' 'electropop' 'emo'
 'entehno' 'epicmetal' 'estrada' 'ethnic' 'eurofolk' 'european'
 'experimental' 'extrememetal' 'fado' 'film' 'fitness' 'flamenco' 'folk'
 'folklore' 'folkmetal' 'folkrock' 'folktronica' 'forró' 'frankreich'
 'französisch' 

Busca en la lista para encontrar duplicados implícitos del género `hiphop`. Estos pueden ser nombres escritos incorrectamente o nombres alternativos para el mismo género.

Verás los siguientes duplicados implícitos:
* `hip`
* `hop`
* `hip-hop`

Para deshacerte de ellos, declara la función `replace_wrong_genres()` con dos parámetros:
* `wrong_genres=` — la lista de duplicados;
* `correct_genre=` — el string con el valor correcto.

La función debería corregir los nombres en la columna `'genre'` de la tabla `df`, es decir, remplaza cada valor de la lista `wrong_genres` con el valor en `correct_genre`. Utiliza un bucle `'for'` para iterar sobre la lista de géneros incorrectos y reemplazarlos con el género correcto en la lista principal.

In [20]:
# función para reemplazar duplicados implícitos
def replace_wrong_genres(df,column,wrong_genres,correct_genre):
    for wrong_genre in wrong_genres:
        df[column]=df[column].replace(wrong_genre,correct_genre)
    return df    


Llama a `replace_wrong_genres()` y pásale argumentos para que retire los duplicados implícitos (`hip`, `hop` y `hip-hop`) y los reemplace por `hiphop`:

In [21]:
# eliminar duplicados implícitos
duplicates=["hip","hop"]
genre="hip-hop"
replace_wrong_genres(df,"genre",duplicates,genre)

user_id                              track            artist  \
0      FFB692EC                  Kamigata To Boots  The Mass Missile   
1      55204538        Delayed Because of Accident  Andreas Rönnberg   
2        20EC38                  Funiculì funiculà       Mario Lanza   
3      A3DD03C9              Dragons in the Sunset        Fire + Ice   
4      E2DC1FAE                        Soul People        Space Echo   
...         ...                                ...               ...   
61248  729CBB09                            My Name            McLean   
61249  D08D4A55  Maybe One Day (feat. Black Spade)       Blu & Exile   
61250  C5E3A0D5                          Jalopiina            unknow   
61251  321D0506                      Freight Train     Chas McDevitt   
61252  3A64EF84          Tell Me Sweet Little Lies      Monica Lopez   

            genre         city      time        day  
0            rock  Shelbyville  20:28:33  Wednesday  
1            rock  Springfield  14:07:09     Friday  
2             pop  Shelbyville  20:58:07  Wednesday  
3            folk  Shelbyville  08:37:09     Monday  
4           dance  Springfield  08:34:34     Monday  
...           ...          ...       ...        ...  
61248         rnb  Springfield  13:32:28  Wednesday  
61249     hip-hop  Shelbyville  10:00:00     Monday  
61250  industrial  Springfield  20:09:26     Friday  
61251        rock  Springfield  21:43:59     Friday  
61252     country  Springfield  21:59:46     Friday  

[61253 rows x 7 columns]

Asegúrate de que los nombres duplicados se hayan eliminado. Muestra la lista de valores únicos de la columna `'genre'` una vez más:

In [23]:
# comprobación de duplicados implícitos
print(df["genre"].sort_values().unique())
print(df["genre"].nunique())


['acid' 'acoustic' 'action' 'adult' 'africa' 'afrikaans' 'alternative'
 'ambient' 'americana' 'animated' 'anime' 'arabesk' 'arabic' 'arena'
 'argentinetango' 'art' 'audiobook' 'avantgarde' 'axé' 'baile' 'balkan'
 'beats' 'bigroom' 'black' 'bluegrass' 'blues' 'bollywood' 'bossa'
 'brazilian' 'breakbeat' 'breaks' 'broadway' 'cantautori' 'cantopop'
 'canzone' 'caribbean' 'caucasian' 'celtic' 'chamber' 'children' 'chill'
 'chinese' 'choral' 'christian' 'christmas' 'classical' 'classicmetal'
 'club' 'colombian' 'comedy' 'conjazz' 'contemporary' 'country' 'cuban'
 'dance' 'dancehall' 'dancepop' 'dark' 'death' 'deep' 'deutschrock'
 'deutschspr' 'dirty' 'disco' 'dnb' 'documentary' 'downbeat' 'downtempo'
 'drum' 'dub' 'dubstep' 'eastern' 'easy' 'electronic' 'electropop' 'emo'
 'entehno' 'epicmetal' 'estrada' 'ethnic' 'eurofolk' 'european'
 'experimental' 'extrememetal' 'fado' 'film' 'fitness' 'flamenco' 'folk'
 'folklore' 'folkmetal' 'folkrock' 'folktronica' 'forró' 'frankreich'
 'französisch' 

<div class="alert alert-block alert-danger">
<b>Comentario del revisor </b> <a class="tocSkip"></a>

~~Esta construyendo la función de manera correcta, sin embaego es importante presentar estos valores únicos de manera ordenada. Te proporcionaré una guía sobre cómo lograrlo.~~

```python
df['genre'].sort_values().unique()
```
    

</div>



<div class="alert alert-block alert-success">
<strong>Comentario del revisor #2</strong> <a class="tocSkip"></a>
    
Buen trabajo.
</div>


[Volver a Contenidos](#back)

### Tus observaciones <a id='data_preprocessing_conclusions'></a>

`Describa brevemente lo que ha observado al analizar los duplicados, cómo abordó sus eliminaciones y qué resultados logró.`
por lo que vi para el analisis es importante que la columna genre este en optimas condiciones para esto elimine los nombres duplicados porque esto podria modificar el resultado final, finalmente obtuve la columna de genre sin errores y listo para el analisis

[Volver a Contenidos](#back)

## Etapa 3. Prueba de hipótesis <a id='hypotheses'></a>

### Hipótesis 1: comparar el comportamiento del usuario en las dos ciudades <a id='activity'></a>

La primera hipótesis afirma que existen diferencias en la forma en que los usuarios y las usuarias de Springfield y Shelbyville consumen música. Para comprobar esto, usa los datos de tres días de la semana: lunes, miércoles y viernes.

* Agrupa a los usuarios y las usuarias por ciudad.
* Compara el número de pistas que cada grupo reprodujo el lunes, el miércoles y el viernes.


Realiza cada cálculo por separado.

El primer paso es evaluar la actividad del usuario en cada ciudad. Agrupa los datos por ciudad y encuentra el número de canciones reproducidas en cada grupo.



In [24]:
# contando las pistas reproducidas en cada ciudad
city = df.groupby("city")["track"].count()
print(city)

city
Shelbyville    18512
Springfield    42741
Name: track, dtype: int64


`Comenta tus observaciones aquí` segun el resultado para la ciudad shelbyville tenemos 18512 pistas reproducidas y para springfield 42741 pistas reproducidas por lo que springfield tienes muchas mas reproducciones que shelbyville

Ahora agrupa los datos por día de la semana y encuentra el número de pistas reproducidas el lunes, el miércoles y el viernes.


In [29]:
# Cálculo de las pistas reproducidas cada día de la semana
day=df.groupby("day")["track"].count()
print(day)


day
Friday       21840
Monday       21354
Wednesday    18059
Name: track, dtype: int64


<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

~~Estas realizando una parte de los agrupamientos de manera correcta, sin embargo no estás seleccionando la variable en la que se ejecuten los cálculos agrupados. Por ende, se genera un valor para cada columna, lo que resulta en información redundante. A continuación, te daré un ejemplo de cómo puedes arreglar el primer agrupamiento.~~
    
```python
df.groupby('city')['track'].count()
```
</div>


<div class="alert alert-block alert-success">
<strong>Comentario del revisor #2</strong> <a class="tocSkip"></a>
    
Buen trabajo.</div>


`Comenta tus observaciones aquí`dentro de los tres dias que tenemos los datos, el dia con mas reproducciones es el viernes con 21840 reproducciones, segundo el lunes con 21354 reproducciones y por ultimo el miercoles con 18059 reproducciones

Ya sabes cómo contar entradas agrupándolas por ciudad o día. Ahora necesitas escribir una función que pueda contar entradas según ambos criterios simultáneamente.

Crea la función `number_tracks()` para calcular el número de canciones reproducidas en un determinado día **y** ciudad. La función debe aceptar dos parámetros:

- `day`: un día de la semana para filtrar. Por ejemplo, `'Monday'`.
- `city`: ciudad: una ciudad para filtrar. Por ejemplo, `'Springfield'`.

Dentro de la función, aplicarás un filtrado consecutivo con indexación lógica.

Primero filtra los datos por día y luego filtra la tabla resultante por ciudad.

Después de filtrar los datos por dos criterios, cuenta el número de valores de la columna 'user_id' en la tabla resultante. Este recuento representa el número de entradas que estás buscando. Guarda el resultado en una nueva variable y devuélvelo desde la función.

In [80]:
def number_tracks(day,city):# <crear la función number_tracks()>
    # declararemos la función con dos parámetros: day=, city=.
    track_list=df[(df["day"]==day) & (df["city"]==city)]# deja que la variable track_list almacene las filas df en las que
    # el valor del nombre de la columna ‘day’ sea igual al parámetro day= y, al mismo tiempo,
    # el valor del nombre de la columna ‘city’ sea igual al parámetro city= (aplica el filtrado consecutivo con indexación lógica)
    track_list_count=len(track_list["user_id"])# deja que la variable track_list_count almacene el número de valores de la columna 'user_id' en track_list
# (igual al número de filas en track_list después de filtrar dos veces).
    return track_list_count# permite que la función devuelva un número: el valor de track_list_count.

# la función cuenta las pistas reproducidas en un cierto día y ciudad.
# primero recupera las filas del día deseado de la tabla,
# después filtra las filas de la ciudad deseada del resultado,
# luego encuentra el número de pistas en la tabla filtrada,
# y devuelve ese número.
# para ver lo que devuelve, envuelve la llamada de la función en print().


# empieza a escribir tu código aquí


Llama a `number_tracks()` seis veces, cambiando los valores de los parámetros, para que recuperes los datos de ambas ciudades para cada uno de los tres días.

In [81]:
print(number_tracks("Monday","Springfield"))# el número de canciones reproducidas en Springfield el lunes


15740


In [82]:
# el número de canciones reproducidas en Shelbyville el lunes
print(number_tracks("Monday","Shelbyville"))

5614


In [83]:
# el número de canciones reproducidas en Springfield el miércoles
print(number_tracks("Wednesday","Springfield"))

11056


In [88]:
# el número de canciones reproducidas en Shelbyville el miércoles
print(number_tracks("Wednesday","Shelbyville"))

7003


In [91]:
# el número de canciones reproducidas en Springfield el viernes
print(number_tracks("Friday","Springfield"))

15945


In [92]:
# el número de canciones reproducidas en Shelbyville el viernes
print(number_tracks("Friday","Shelbyville"))

5895


Utiliza `pd.DataFrame` para crear una tabla, donde
* los encabezados de la tabla son: `['city', 'monday', 'wednesday', 'friday']`
* Los datos son los resultados que conseguiste de `number_tracks()`

In [94]:
# tabla con los resultados
import pandas as pd
data=[["Springfield",number_tracks("Monday","Springfield"),number_tracks("Wednesday","Springfield"),number_tracks("Friday","Springfield")],["Shelbyville",number_tracks("Monday","Shelbyville"),number_tracks("Wednesday","Shelbyville"),number_tracks("Friday","Shelbyville")]]
columns = ['City','monday', 'wednesday', 'friday']  
print(pd.DataFrame(data,columns=columns))


          City  monday  wednesday  friday
0  Springfield   15740      11056   15945
1  Shelbyville    5614       7003    5895


<div class="alert alert-block alert-success">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>

¡Genial! Buen trabajo con la función `number_tracks()` y la tabla de resultados.
</div>


**Conclusiones**

`Comente si la primera hipótesis es correcta o debe rechazarse. Explicar tu razonamiento`La primera hipotesis es correcta porque segun la utlima tabla hay diferencias entre el comportamiento de los usuarios en Spriengfield y Shelbyville, siendo Sprinegfield donde se encuentra un mayor numero de reproducciones de musica en los tres dias clave

[Volver a Contenidos](#back)

### Hipótesis 2: música al principio y al final de la semana <a id='week'></a>

Según la segunda hipótesis, el lunes por la mañana y el viernes por la noche, los ciudadanos de Springfield escuchan géneros que difieren de los que disfrutan los usuarios de Shelbyville.

Cree dos tablas con los nombres proporcionados en los dos bloques de código a continuación:
* Para Springfield — `spr_general`
* Para Shelbyville — `shel_general`

In [31]:
# crear la tabla spr_general a partir de las filas df
# donde los valores en la columna 'city' es 'Springfield'


columns=['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day']
spr_general=pd.DataFrame(df[df["city"]=="Springfield"],columns=columns)


print(spr_general)


        user_id                          track                   artist  \
1      55204538    Delayed Because of Accident         Andreas Rönnberg   
4      E2DC1FAE                    Soul People               Space Echo   
6      4CB90AA5                           True             Roman Messer   
7      F03E1C1F               Feeling This Way          Polina Griffith   
8      8FA1D3BE                       L’estate              Julia Dalia   
...         ...                            ...                      ...   
61247  83A474E7  I Worship Only What You Bleed  The Black Dahlia Murder   
61248  729CBB09                        My Name                   McLean   
61250  C5E3A0D5                      Jalopiina                   unknow   
61251  321D0506                  Freight Train            Chas McDevitt   
61252  3A64EF84      Tell Me Sweet Little Lies             Monica Lopez   

              genre         city      time        day  
1              rock  Springfield  14:07:09 

In [32]:
# crear la tabla shel_general a partir de las filas df
# donde los valores en la columna 'city' es 'Shelbyville'
columns=['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day']

shel_general=pd.DataFrame(df[df["city"]=="Shelbyville"],columns=columns)
print(shel_general)


        user_id                              track              artist  \
0      FFB692EC                  Kamigata To Boots    The Mass Missile   
2        20EC38                  Funiculì funiculà         Mario Lanza   
3      A3DD03C9              Dragons in the Sunset          Fire + Ice   
5      842029A1                             Chains            Obladaet   
9      E772D5C0                          Pessimist              unknow   
...         ...                                ...                 ...   
61239  D94F810B        Theme from the Walking Dead  Proyecto Halloween   
61240  BC8EC5CF       Red Lips: Gta (Rover Rework)               Rover   
61241  29E04611                       Bre Petrunko       Perunika Trio   
61242  1B91C621             (Hello) Cloud Mountain     sleepmakeswaves   
61249  D08D4A55  Maybe One Day (feat. Black Spade)         Blu & Exile   

            genre         city      time        day  
0            rock  Shelbyville  20:28:33  Wednesday  
2  

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Hay algunos errores en esta sección, primero que todo recuerda que las librerías solo se importan al principio del proyecto. Por otro lado, el código está repetido, deberías fijarte y dejar ordenado de mejor manera la sección.</div>

<div class="alert alert-block alert-success">
<strong>Comentario del revisor #2</strong> <a class="tocSkip"></a>
    
Buen trabajo.
</div>


Escribe la función `genre_weekday()` con cuatro parámetros:
* Una tabla para los datos (`df`)
* El día de la semana (`day`)
* La marca de fecha y hora en formato 'hh:mm:ss' (`time1`)
* La marca de fecha y hora en formato 'hh:mm:ss' (`time2`)

La función debe devolver los 15 géneros más populares en un día específico dentro del período definido por las dos marcas de tiempo, junto con sus respectivos recuentos de reproducción.
Aplica la misma lógica de filtrado consecutiva, pero usa cuatro filtros esta vez y luego crea una nueva columna con los recuentos de reproducción respectivos.
Ordena el resultado de un recuento más grande a uno más pequeño y devuélvelo.

In [174]:
# 1) Deja que la variable genre_df almacene las filas que cumplen varias condiciones:
#    - el valor de la columna 'day' es igual al valor del argumento day=
#    - el valor de la columna 'time' es mayor que el valor del argumento time1=
#    - el valor en la columna 'time' es menor que el valor del argumento time2=
#    Utiliza un filtrado consecutivo con indexación lógica.
  

# 2) Agrupa genre_df por la columna 'genre', toma una de sus columnas,
#    y utiliza el método size() para encontrar el número de entradas por cada uno de
#    los géneros representados; almacena los Series resultantes en
#    la variable genre_df_count

# 3) Ordena genre_df_count en orden descendente de frecuencia y guarda el resultado
#    en la variable genre_df_sorted

# 4) Devuelve un objeto Series con los primeros 15 valores de genre_df_sorted - los 15
#    géneros más populares (en un determinado día, en un determinado periodo de tiempo)

# escribe tu función aquí

    # filtrado consecutivo
    # Crea la variable genre_df que almacenará solo aquellas filas df donde el día es igual a day=
    # escribe tu código aquí

    # Filtra genre_df nuevamente para almacenar solo las filas donde el tiempo es menor que time2=
     # escribe tu código aquí

    # Filtra genre_df una vez más para almacenar solo las filas donde el tiempo es mayor que time1=
     # escribe tu código aquí

    # Agrupa el DataFrame filtrado por la columna con los nombres de los géneros, selecciona la columna 'genre',
    # y encuentra el número de filas para cada género con el método count()
def genre_weekday(df,day,time1,time2):
    genre_df=df[(df["day"]==day)&(df["time"]>time1)&(df["time"]<time2)]
    genre_df=genre_df["genre"]# escribe tu código aquí
    genre_df_count=genre_df.value_counts()

    # Ordenaremos el resultado en orden descendente (por lo que los géneros más populares aparecerán primero en el objeto Series)
    genre_df_sorted = genre_df_count.sort_values(ascending=False)# escribe tu código aquí

    # Devuelve un objeto de Series con los primeros 15 valores de genre_df_sorted: los 15 géneros más populares (en un día determinado, dentro de un período de timeframe)
    return genre_df_sorted[:15]

Compara los resultados de la función `genre_weekday()` para Springfield y Shelbyville el lunes por la mañana (de 7 a 11) y el viernes por la tarde (de 17:00 a 23:00). Utiliza el mismo formato de hora de 24 horas que el conjunto de datos (por ejemplo, 05:00 = 17:00:00):

In [175]:
# llamando a la función para el lunes por la mañana en Springfield (utilizando spr_general en vez de la tabla df)
genre_weekday(spr_general,"Monday","07:00:00","11:00:00")

pop            781
dance          549
electronic     480
rock           474
hip-hop        282
ruspop         186
world          181
rusrap         175
alternative    164
unknow         161
classical      157
metal          120
jazz           100
folk            97
soundtrack      95
Name: genre, dtype: int64

In [176]:
# llamando a la función para el lunes por la mañana en Shelbyville (utilizando shel_general en vez de la tabla df)
genre_weekday(shel_general,"Monday","07:00:00","11:00:00")

pop            218
dance          182
rock           162
electronic     147
hip-hop         79
ruspop          64
alternative     58
rusrap          55
jazz            44
classical       40
world           36
rap             32
soundtrack      31
metal           27
rnb             27
Name: genre, dtype: int64

In [178]:
# llamando a la función para el viernes por la tarde en Springfield
genre_weekday(spr_general,"Friday","17:00:00","23:00:00")

pop            713
rock           517
dance          495
electronic     482
hip-hop        267
world          208
ruspop         170
classical      163
alternative    163
rusrap         142
jazz           111
unknow         110
soundtrack     105
rnb             90
metal           88
Name: genre, dtype: int64

In [179]:
# llamando a la función para el viernes por la tarde en Shelbyville
genre_weekday(shel_general,"Friday","17:00:00","23:00:00")

pop            256
rock           216
electronic     216
dance          210
hip-hop         94
alternative     63
jazz            61
classical       60
rusrap          59
world           54
ruspop          47
unknow          47
soundtrack      40
metal           39
rap             36
Name: genre, dtype: int64

**Conclusiones**

`Comente si la segunda hipótesis es correcta o debe rechazarse. Explica tu razonamiento.` con respecto a los dias lunes por la mañana y el viernes por la noche se puede observar que hay una similitud en los generos que se escuchan en Shelbyville y Springfield , por lo tanto la segunda hipotesis es parcialmente correcta

<div class="alert alert-block alert-success">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>

Buen trabajo con la función.
</div>

[Volver a Contenidos](#back)

### Hipótesis 3: preferencias de género en Springfield y Shelbyville <a id='genre'></a>

Hipótesis: Shelbyville ama la música rap. A los residentes de Springfield les gusta más el pop.

Agrupa la tabla `spr_general` por género y encuentra el número de canciones reproducidas de cada género con el método `count()`. Luego ordena el resultado en orden descendente y guárdalo en la variable `spr_genres`.

In [28]:
# escribe una línea de código que:
# 1. agrupe la tabla spr_general por la columna 'genre';
# 2. cuente los valores 'genre' con count() en la agrupación;
# 3. ordene el Series resultante en orden descendente y lo guarde en spr_genres.

spr_genre=spr_general["genre"].value_counts()
spr_genres=spr_genre.sort_values(ascending=False)
print(spr_genres)

pop               5892
dance             4435
rock              3965
electronic        3786
hip-hop           2043
                  ... 
showtunes            1
malaysian            1
piano                1
sängerportrait       1
axé                  1
Name: genre, Length: 251, dtype: int64


Imprime las 10 primeras filas de `spr_genres`:

In [29]:
# muestra los primeros 10 valores de spr_genres
print(spr_genres[:10])

pop            5892
dance          4435
rock           3965
electronic     3786
hip-hop        2043
classical      1616
world          1432
alternative    1379
ruspop         1372
rusrap         1161
Name: genre, dtype: int64


Ahora haz lo mismo con los datos de Shelbyville.

Agrupa la tabla `shel_general` por género y encuentra el número de canciones reproducidas de cada género. Después, ordena el resultado en orden descendente y guárdalo en la tabla `shel_genres`:


In [34]:
# escribi una línea de código que:
# 1. agrupe la tabla shel_general por la columna 'genre';
# 2. cuente los valores 'genre' con count() en la agrupación;
# 3. ordene el Series resultante en orden descendente y lo guarde en shel_genres.
shel_genre=shel_general["genre"].value_counts()
shel_genres=shel_genre.sort_values(ascending=False)
print(shel_genres)

pop            2431
dance          1932
rock           1879
electronic     1736
hip-hop         934
               ... 
rhythm            1
garage            1
orchestral        1
traditional       1
japanese          1
Name: genre, Length: 203, dtype: int64


Imprime las 10 primeras filas de `shel_genres`:

In [35]:
# imprimir las 10 primeras filas de shel_genres
print(shel_genres[:10])

pop            2431
dance          1932
rock           1879
electronic     1736
hip-hop         934
alternative     649
classical       646
rusrap          564
ruspop          538
world           515
Name: genre, dtype: int64


**Conclusión**

`Comente si la tercera hipótesis es correcta o debe rechazarse. Explica tu razonamiento.`la tercera hipotesis es parcialmente correcta porque la musica rap no figura dentro de los generos mas escuchados en shelbyville pero se acepta parcialmente porque en springfield el genero mas escuchado avalado por la hipotesis es pop

[Volver a Contenidos](#back)

# Conclusiones <a id='end'></a>

`Resuma sus conclusiones sobre cada hipótesis aquí` con respecto a la primera hipotesis en los dias lunes, miercoles y viernes es correcta porque segun los datos hay diferencias en el comportamiento de las habitantes de Shelbyville y Springfield.
En la segunda hipotesis es parcialmente correcta porque en la comparacion de generos mas escuchados en las dos ciudades hay generos que aparecen en las dos ciudades y otras que no.
En la tercera hipotesis se planteo dos condiciones y se cumplio solo una donde en spriengfield es pop el genero mas escuchado

### Nota
En proyectos de investigación reales, la prueba de hipótesis estadística es más precisa y cuantitativa. También ten en cuenta que no siempre se pueden sacar conclusiones sobre una ciudad entera a partir de datos de una sola fuente.

Aprenderás más sobre la prueba de hipótesis en el sprint de análisis estadístico de datos.

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
# Comentario General
    
~~Te felicito por el desarrollo del momento hasta el momento, ahora bien, debes tener en cuenta los comentarios que he generado para la próxima entrega.~~


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

# Comentario General #2

Buen trabajo, Yair. Has culminado el proyecto.
</div>


[Volver a Contenidos](#back)